### ISSUE TITLE
Incorrect picture size on RPi4 in KMSDRM mode

### ISSUE THREAD

**Issue by boras-pl** [libsdl-org/SDL#4417]
I noticed that https://github.com/libsdl-org/SDL/commit/0219928dd5416be9dfb21217c36f4d1b794e4b80 introduces a regress that causes cut and moved video on at least two projects:
- https://github.com/midwan/amiberry developed by @midwan (compiled from sources)
- chocolate-doom (installed via apt)

The latest revision behaves in the same way. As I am not a developer of any SDL2 software I can not say if they initialize video mode correctly or not. I just observe results only.

My environment:
- RaspiOS 64-bit lite for Raspberry Pi 4b (no X11, pure console) - https://downloads.raspberrypi.org/raspios_lite_arm64/images/raspios_lite_arm64-2021-05-28/2021-05-07-raspios-buster-arm64-lite.zip 
- pi@raspberrypi:~/SDL $ uname -a
Linux raspberrypi 5.10.17-v8+ #1421 SMP PREEMPT Thu May 27 14:01:37 BST 2021 aarch64 GNU/Linux
- SDL2 compiled manually
````
./configure --enable-video-kmsdrm --enable-video-opengl --enable-video-opengles --disable-video-wayland --disable-video-rpi --disable-video-x11 --disable-video-vivante --disable-video-cocoa --disable-video-directfb --disable-video-vulkan --disable-directx
````
![IMG_20210604_143118](https://user-images.githubusercontent.com/59491904/120803107-816fca80-c543-11eb-9703-389866043f9a.jpg)
![IMG_20210604_143136](https://user-images.githubusercontent.com/59491904/120803111-82a0f780-c543-11eb-91c3-1af94a3d809d.jpg)

**Comment by Cacodemon345** [comment]
I think the `SDL_WINDOWSEVENT_RESIZED` event should be sent to the application on setting video modes on KMSDRM.

**Comment by vanfanel** [comment]
KMSDRM is now broken everywhere.
I left it working perfectly well last time I touched the code.

These commits (NOT from me) are to blame. Feel free to join the discussion:
https://github.com/libsdl-org/SDL/pull/4396

**Comment by inolen** [comment]
@boras-pl is it exactly that commit which causes the regression?

If so, can you try restoring the second SDL_WINDOWEVENT_RESIZED event as seen here:
https://github.com/libsdl-org/SDL/blob/99ef03b96ec95c4fb044c600e48b708b6f2fe83e/src/video/kmsdrm/SDL_kmsdrmvideo.c#L1359

and see if that resolves the issue? Perhaps there's a dependency on that being called before KMSDRM_CreateSurfaces in some cases.

**Comment by boras-pl** [comment]
Ok, I'll look at this in the evening

**Comment by inolen** [comment]
Since you have a repro on this, if you have the time / want to real time debug, feel free to hit me up on Discord (https://discord.gg/zuEAVTG) and I can try to help.

**Comment by boras-pl** [comment]
I have checked 56e96254c23105d7c7657f32d50661435d838b6a out and below are my thoughts:

1. Both Amiberry and chocolate-doom go into
https://github.com/libsdl-org/SDL/blob/56e96254c23105d7c7657f32d50661435d838b6a/src/video/kmsdrm/SDL_kmsdrmvideo.c#L1317
so recreation the event in your place does nothing.
2. There is no windata in the function. I can not simply reuse it from 99ef03b96ec95c4fb044c600e48b708b6f2fe83e because of:
````
error: 'SDL_WindowData' {aka 'struct SDL_WindowData'} has no member named 'surface_w'
         windata->surface_w = dispdata->original_mode.hdisplay;
````
3. Another blind guess:
````
    if ((window->flags & SDL_WINDOW_FULLSCREEN) ==
                         SDL_WINDOW_FULLSCREEN)
    {
        /* Tell app about the size we have determined for the window,
           so SDL pre-scales to that size for us. */
        printf("I am here\n");
        SDL_SendWindowEvent(window, SDL_WINDOWEVENT_RESIZED,
                        dispdata->mode.hdisplay, dispdata->mode.vdisplay);
    }
````
prints _I am here_ but the problem still exists.

Unfortunately my time is very limited these days for reasons beyond my control. However, I am able to do some quick tests almost every evening. Proof of concepts that I could compile are welcome.

**Comment by inolen** [comment]
Thanks for that info! I think I understand the issue now.

I believe what's happening is that the default CRTC found for your display is tiny, which makes sense why I can't repro this; it's an issue that's display specific.

When the display is first added, the default CRTC is used to provide the initial display mode that's used when toggling fullscreen (this mode is only overridden by a call to KMSDRM_SetDisplayMode). This code previously used to work because calling KMSDRM_SetWindowFullscreen would search for a mode similar to the window size (overriding anything previously set by KMSDRM_SetDisplayMode unfortunately).

There's only two options I see here:

1. Try to figure out why the default CRTC is so small (not sure if that's a fool's errand). It'd be great if we always got a mode that was similar to the vt you're launching from, but I don't know enough about these APIs to know how realistic that is.
2. Make KMSDRM_CreateWindow use the window size to update the display's fullscreen mode with a size that's appropriate for the window. Given this backend only deals with single _visible_ windows (it supports multiple windows for multiple contexts, but only a single visible window), I think this is likely fine. The worst case bug I can think of is the secondary context causes the default fullscreen mode to be incorrect. We could only use the newly selected mode if it's bigger than the current default mode (so at least things are always visible and never clipped). I'm not sure of a better solution if we can't accurately get a sane default mode.

@icculus any ideas on that?

@boras-pl if you can, try compiling https://github.com/inolen/SDL/commits/kmsdrm_regressions. I've pushed a commit there that partially implements the 2nd idea there.

**Comment by boras-pl** [comment]
Hello.

I've found few minutes to check it:
https://github.com/inolen/SDL/commit/5d00ba84282af6e92d5131d3f310d175bb363bdd

Amiberry displays it in the same way regardless of video settings (window, fullscreen, full window). I don't know how it is implemented there.
![IMG_20210610_071602](https://user-images.githubusercontent.com/59491904/121471231-98893f00-c9bf-11eb-82d2-8e2751f74d2f.jpg)
![IMG_20210610_071319](https://user-images.githubusercontent.com/59491904/121471270-a76ff180-c9bf-11eb-9c9f-ae59cd890825.jpg)

**Comment by midwan** [comment]
The code Amiberry uses to switch between Full Screen, Full-Window and Windowed mode is here: https://github.com/midwan/amiberry/blob/54dfe34953ec1a24f1f4afd0d3eae4d72de6aead/src/osdep/amiberry_gfx.cpp#L562

**Comment by Cacodemon345** [comment]
Seems like Amiberry sets the window to fullscreen mode using the desktop mode resolution when KMSDRM backend is used.

**Comment by inolen** [comment]
I think I see one more issue on the KMSDRM side I'll push a wip fix for tonight to that branch. However, @midwan  I'm curious what's the intent of https://github.com/midwan/amiberry/blob/54dfe34953ec1a24f1f4afd0d3eae4d72de6aead/src/osdep/amiberry_gfx.cpp#L583

Shouldn't that be using SDL_SetWindowDisplayMode when switching to actual fullscreen?

**Comment by midwan** [comment]
It's possible that my implementation is not entirely correct, so I'm open for suggestions. This used to work, so I've left it like that... And it does work as expected under X11

**Comment by inolen** [comment]
@midwan it's more likely that I'm just not familiar with all the nuances of the SDL code here. Looking now, it seems that the backend-agnostic SDL code will actually try and already do the same logic the KMSDRM backend is doing (and find an appropriate mode to set based on the window size) if no fullscreen mode has ever been set by SDL_SetWindowDisplayMode when SDL_SetWindowSize is called, so that seems good actually.

**Comment by boras-pl** [comment]
@midwan I have checked what is going on in:
https://github.com/midwan/amiberry/blob/54dfe34953ec1a24f1f4afd0d3eae4d72de6aead/src/osdep/amiberry_gfx.cpp#L572
an I have interesting observations. For all possible scenarios:
- fullscreen 800x600
- fullscreen 1920x1080 (native for my display)
- full-window
- window
I try to print some values:
````
write_log("window_flags=%d\n", window_flags);
write_log("is_fullwindow=%d\n", is_fullwindow);
write_log("is_fullscreen=%d\n", is_fullscreen);
````
I always get the same results:
````
window_flags=5639
is_fullwindow=1
is_fullscreen=1
````
Thus SDL_SetWindowFullscreen is not called in any above scenarios.

@inolen does SDL_GetWindowFlags work correctly?

Do not hesitate to correct me if I messed something.

**Comment by boras-pl** [comment]
Things are more complicated as I thought:
- my Acer Predator (native 1920x1080) has two inputs: Display Port and HDMI. Any resolution on DP is always upscaled by the display to the native resolution. But I use it with my PC only. Anything on HDMI is displayed as is without upscaling.
- attached picture from Amiberry was from Raspberry PI4 connected to HDMI. In this case  the 800x600 mode is set. I can check it in the monitor's menu.
- when I hardcode 
````
is_fullwindow=0;
is_fullscreen=0; 
````
Amiberry finally executes desired code (if fullscreen mode is chosen) when it starts an emulator:
````
SDL_SetWindowFullscreen(mon-&gt;sdl_window, SDL_WINDOW_FULLSCREEN);
SDL_SetWindowSize(mon->sdl_window, p->gfx_monitor[0].gfx_size_fs.width, p->gfx_monitor[0].gfx_size_fs.height);
````      
Then it works more or less correctly. There are sometimes some strange side-effects like blinking or black pictures, periodic lack of mouse pointer, etc. But it usually works and reported resolution (I tested a few) is correct. It does not matter that it is displayed in a small window when <1920x1080. It is a feature of my HDMI input port of display as I written above.

- I still don't know where the value of SDL_GetWindowFlags comes from.

It does not look so bad, but there are still some hard to test and describe corner-cases. You should be aware of existence of many incorrectly written SDL applications that will stop working even when your code will be 100% in accordance to the specification (if any exists). As I am not a SDL developer I may mislead you unintentionally.

**Comment by inolen** [comment]
Hi @boras-pl sorry I've been busy this weekend. I pushed a change Friday, did that make any difference for you or are you still seeing the regression?

**Comment by vanfanel** [comment]
@inolen I have cloned and built your kmsdrm_regression branch.
-Programs don't work fullscreen. Setting any 2D SDL2 program fullscreen does NOT scale the program fullscreen.
-OpenGL programs (sm64ex) work fullscreen, but changing from fullscreen to windowed mode does nothing.

As I initially said, **all this worked fine before your commits.**
Allow me to insist: roll back all those commits, and take your time to test the changes with different applications before comitting again. There's no reason to break the backend.

**Comment by vanfanel** [comment]
@inolen I have cloned and built your **kmsdrm_regression** branch.
-Programs don't work fullscreen. Setting any 2D SDL2 program fullscreen does NOT scale the program fullscreen.
-OpenGL programs (sm64ex) work fullscreen, but changing from fullscreen to windowed mode does nothing.

As I initially said, **all this worked fine before your commits.**
Allow me to insist: roll back all those commits, and take your time to test the changes with different applications before comitting again. There's no reason to break the backend.

**Comment by inolen** [comment]
Hi @vanfanel, please stop going on about how the sky is falling and requesting for a rollback or nothing; this isn't useful. If you want to help move forward in a positive direction, help debug given that you can actually reproduce the issue.

I'll see if I can reproduce the issue on amiberry tonight, and if I can't I'll spend some cycles trying to think about what could be going on with the info you've provided @boras-pl.

**Comment by boras-pl** [comment]
I'll check it in the evening.

**Comment by vanfanel** [comment]
@inolen Ok, ok. I'll simply keep building and reporting. That's the only thing I can do right now to help.

**Comment by boras-pl** [comment]
guys,

anything below is my guess only:

SDL_GetWindowFlags works ok. At least in the same way as old SDL revisions.
As I correctly understand Amiberry creates a main windows 800x600 SDL_WINDOW_FULLSCREEN_DESKTOP = ( SDL_WINDOW_FULLSCREEN | 0x00001000 )
Before your change in SDL a video mode had not been changed and everything  was upscaled to the current VT console resolution.
With the latest kmsdrm_regression SDL it creates a real 800x600 mode. Then it checks if the current mode is SDL_WINDOW_FULLSCREEN and does nothing. That's why it cannot set another fullscreen resolution. So, this particular behaviour is specific for Amiberry only.
Any tests do not make sense until fine-tuning of Amiberry logic is done. However I don't yet understand how it works in X11.

As I understand @inolen  wants to have a mode setting feature (which seams to be tempting, but it may break compatibility with many s/w. @vanfanel  wants to have everything upscaled to one resolution.
Both of you have own rights and I'm afraid that I am not so competent to settle a dispute.

One more thing, I guess that SDL_WINDOW_FULLSCREEN_DESKTOP should not change the current console screen mode. If yes, it is a problem in kmsdrm_regression as well.

Too many guesses, too many ideas. Are there any simple test cases in the world?

**Comment by inolen** [comment]
Hi @boras-pl,

I'll be looking at this soon but to be clear - supporting SDL's mode setting API for fullscreen windows and allowing for automatically finding a mode to set for non-fullscreen windows is entirely possible. There's just some debugging involved it seems and it's unfortunate I haven't been able to repro the issues experienced thus far.

**Comment by slime73** [comment]
For comparison's sake, here's how window sizing with respect to fullscreen options works on other platforms (and therefore how I'd expect it to work when the KMSDRM backend is used) – I believe all of what I'm writing is accurate but please correct me if I'm wrong:

* Creating a window without any fullscreen flags will attempt to make the window at the requested size. If it can't do it, the closest appropriate size may be chosen instead. A windowed-mode window does not change the active display mode. A couple examples of that internal resizing behaviour are:
    • On macOS if the window is resizable and it would cover the dock area, the OS shrinks the size to fit.
    • On iOS, there's no concept of floating app-specified window bounds, so the size request is used as an orientation (landscape/portrait) hint and SDL's code adjusts the true size of the window to fill the screen. This is probably close to what I'd expect on other platforms that don't support app-created floating windows (aside from the orientation part) - such as KMSDRM, I suppose. On iOS, the OS menu bar is left visible when windowed mode is used.

* Creating a window with `SDL_WINDOW_FULLSCREEN_DESKTOP` (or switching a window into that state via `SDL_SetWindowFullscreen`) changes the window's dimensions to fill the monitor's dimensions based on the existing display mode that was active before SDL touched it. It's designed to avoid using a custom display mode, although transitioning from exclusive fullscreen to fullscreen-desktop may change the active display mode if exclusive fullscreen had set a custom one. In other words, it always makes sure the display mode retrieved via `SDL_GetDesktopDisplayMode` is active.

* Exclusive fullscreen (`SDL_WINDOW_FULLSCREEN`) finds the closest match to the requested display mode parameters and changes the monitor's active display mode, while also making the window cover the entire screen at that resolution.

* `SDL_SetWindowDisplayMode` only affects windows that have `SDL_WINDOW_FULLSCREEN` active (exclusive fullscreen). If that window isn't in exclusive fullscreen mode, then it doesn't change the display's active mode when it's called. But making that window exclusive fullscreen via `SDL_SetWindowFullscreen` after calling `SDL_SetWindowDisplayMode` will change the active display mode to the one previously set in SetWindowDisplayMode.

**Comment by vanfanel** [comment]
@slime73 I understand you are right on all that.

**Comment by inolen** [comment]
> @slime73 You are right on all that. But the backend is still broken.
> Some 3D games are also broken, like OpenLara:
> https://github.com/XProger/OpenLara
> 
> OpenLara was of course working fine before the commits by @inolen
> In the current situation, the game won't show any graphics, staying on the TTY console forever.

Please, just stop. No, what @slime73 was not working previously. As an example, SDL_SetWindowFullscreen previously overrode the mode set by a previous SDL_SetWindowDisplayMode. This wasn't correct no matter how much you sit on here trying to complain and rewrite history.

@slime73 this is my understanding as well. I'd submitted some initial commits (which is how we got here) that restored the functionality expected when using SDL_WINDOW_FULLSCREEN / SDL_SetWindowDisplayMode / SDL_SetWindowFullscreen. These changes still didn't address that the KMSDRM backend was modifying the mode requested in SetDisplayMode in other cases (e.g. if you called SetDisplayMode -> SetWindowSize it would override the previous mode). I've tried to further modify this "make a windowed app set the closest mode" functionality with the last two commits in my branch at https://github.com/inolen/SDL/tree/kmsdrm_regressions to avoid ever overriding the requested mode, but I also share your concerns about setting the display mode automatically like this. There's state tracking in the higher-level platform-agnostic code that I'm worried could be going off the rails by doing this.

However, I don't know if these concerns are what's causing this regression here given I can't repro it on my setup. It feels like it has something to do with the default CRT being selected for certain devices.

@midwan is there a way to force amiberry to start in different window modes from the command line to try and repro this? I tried changing the mode in the UI and navigating to "Quit" and relaunching, but it seems to not save the settings when quitting (possibly related to me getting a segfault while shutting down).

**Comment by inolen** [comment]
@boras-pl I tossed together some changes to achieve what I mentioned in my last comment, could you try the latest commit on https://github.com/inolen/SDL/commits/kmsdrm_regressions.

**Comment by vanfanel** [comment]
@inolen Stop what? I just confirmed that what @slime73 said was correct, and reported that OpenLara is broken. What's wrong with that?

I will also test your changes and report back (even if you want me out of the conversation, it seems).

**Comment by boras-pl** [comment]
commit 897c1c0f071f80c2462be3a54392ba6d2f9e5575 (HEAD -> kmsdrm_regressions, origin/kmsdrm_regressions)
amiberry and chocolate doom with their default settings seem to behave as the old well known SDL code. From my point of view it is good enough. Video is upscaled and it maintains the original console resolution. According to your enhancements:  Amiberry still can't switch an emulator to an exclusive fullscreen mode, but I believe it is rather a problem in Amiberry itself than a bug in SDL. When I work the issue around I can force the program to switch to SDL_WINDOW_FULLSCREEN and my display reports that the resolution is changed.
TL;DR
Looks good for me, but I can't check all cases described by @slime73

**Comment by vanfanel** [comment]
@inolen

```
manuel@hp15db0:~/src/sssss/SDL$ git rev-parse --short HEAD
787e6f4b

```

-SDLPop with fullscreen ON isn't filling the screen height as it did before the commits. **You can see how it's intended to work by using the X11 backend, or reverting the commits.**

-pt2-clone with FULLSCREEN=TRUE and FULLSCREENSTRETCH=TRUE isn't filling the whole screen is it did before the commits. 
**You can see how it's intended to work by using the X11 backend, or reverting the commits.**

-OpenLara (https://github.com/XProger/OpenLara)  isn't able to display anything. Since SDL2 KMSDRM takes over the TTY, a system reboot is needed.
**You can see how it's intended to work by using the X11 backend, or reverting the commits.**

etc etc

**Comment by inolen** [comment]
@vanfanel sorry, what version are you testing with exactly?

```pi@raspberrypi:~/projects/sdl/build $  git rev-parse --short HEAD
897c1c0f
```

that's the head of  https://github.com/inolen/SDL/commits/kmsdrm_regressions for me and @boras-pl

**Comment by inolen** [comment]
Just spent the time to setup OpenLara - it was also working. @vanfanel are you sure you're testing my branch correctly? I wasn't able to reproduce a hang on startup with OpenLara even when rolling back to what's been merged in SDL. The only way I was able to make it hang was when using my system's SDL2 which is at 2.0.9 it seems.

**Comment by inolen** [comment]
Just rebased those changes on top of the current SDL's master branch and force pushed again, new head commit is ```git rev-parse --short HEAD
9bbf0bbf
```

Tested again with SDLPoP, amibberry and OpenLara.

**Comment by vanfanel** [comment]
This seems to be fixed now via https://github.com/libsdl-org/SDL/pull/4468

Very good work in the end. Sorry for my earlier attitude.
